src/sonic-framework/rebootbackend/init_thread.h

```diff
@@ -0,0 +1,263 @@
+    m_status.thread_state.mutable_status()->set_status(
+        gnoi::system::RebootStatus_Status::RebootStatus_Status_STATUS_UNKNOWN);
+    m_status.thread_state.mutable_status()->set_message("");
+  }
+
+  bool get_active(void) {
+    const std::lock_guard<std::mutex> lock(m_mutex);
+    return m_status.thread_state.active();
+  }
+
+  void set_detailed_thread_status(ThreadStatus new_status) {
+    const std::lock_guard<std::mutex> lock(m_mutex);
+    if (m_status.thread_state.active()) {
+      m_status.detailed_thread_status = new_status;
+    }
+  }
+
+  void set_success() {
+    const std::lock_guard<std::mutex> lock(m_mutex);
+    if (m_status.thread_state.active()) {
+      m_status.detailed_thread_status = ThreadStatus::DONE;
+      m_status.thread_state.mutable_status()->set_status(
+          gnoi::system::RebootStatus_Status::
+              RebootStatus_Status_STATUS_SUCCESS);
+    }
+  }
+
+  void set_error(ErrorCondition error_condition,
+                 const std::string &error_message) {
+    const std::lock_guard<std::mutex> lock(m_mutex);
+    if (m_status.thread_state.active()) {
+      m_status.detailed_thread_status = ThreadStatus::ERROR;
+      m_status.detailed_thread_error_condition = error_condition;
+      m_status.thread_state.mutable_status()->set_status(
+          gnoi::system::RebootStatus_Status::
+              RebootStatus_Status_STATUS_FAILURE);
+      m_status.thread_state.mutable_status()->set_message(error_message);
+    }
+  }
+
+  void set_inactive() {
+    const std::lock_guard<std::mutex> lock(m_mutex);
+    m_status.thread_state.set_active(false);
+  }
+
+  DetailedStatus get_detailed_thread_status() {
+    const std::lock_guard<std::mutex> lock(m_mutex);
+    return m_status;
+  }
+
+  gnoi::system::RebootStatusResponse get_response() {
+    const std::lock_guard<std::mutex> lock(m_mutex);
+    return m_status.thread_state;
+  }
+
+ private:
+  std::mutex m_mutex;
+  DetailedStatus m_status;
+};
+
+class InitThread {
+ public:
+  InitThread(
+//	     CriticalStateInterface &critical_interface,
+             TelemetryInterface &telemetry_interface,
+             swss::SelectableEvent &m_finished,
+             swss::SelectableEvent &m_stack_unfrozen);
+  virtual ~InitThread() = default;
+
+  // Starts running the init thread tasks. Returns SWSS_RC_SUCCESS if the
+  // internal thread was started successfully, and an error otherwise. If an
+  // error is returned, this call may safely be retried, but will likely
+  // continue to return errors.
+  virtual swss::StatusCode Start();
+
+  // Request InitThread stop/exit. Notifies the internal thread that it must
+  // exit. Only used when platform is shutting down all containers/processes.
+  virtual void Stop(void);
+
+  // Must be called by launching task after notification is sent to m_finished.
+  virtual bool Join(void);
+
+  // Return Status of last reboot attempt.
+  virtual gnoi::system::RebootStatusResponse GetResponse();
```

Comment:
Have we tested RebootStatusResponse ? John, any inputs on how to test this ?

Reply:
I don't have good suggestions here: I'm unfamiliar with availble options.